src/main/cpp/src/zorder.cu

```diff
@@ -27,6 +27,118 @@
 #include <thrust/iterator/constant_iterator.h>
 #include <thrust/iterator/counting_iterator.h>
 
+namespace {
+
+// pretends to be an array of uint32_t, but really only stores
+// the data in a long with a set number of bits allocated for
+// each item (num_bits_per_entry)
+struct long_backed_array {
+  long_backed_array() = delete;
```

Comment:
If we declare `data_type` inside this struct, we should call `long_backed_array::data_type` all the time. Otherwise, we should put it somewhere else (probably above this struct).

Reply:
That is the gist of the change. I was so tempted to create ```
template <typename T>
struct backed_array {
}```

I worry I'm too far in the weeds though and missing the point.